servant-auth-server/src/Servant/Auth/Server/Internal/Class.hs

```diff
@@ -1,49 +1,99 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
 {-# LANGUAGE UndecidableInstances #-}
 module Servant.Auth.Server.Internal.Class where
 
-import Servant.Auth
-import Data.Monoid
+import Data.Proxy
+import GHC.Generics (Generic)
+import Network.Wai (Request)
 import Servant hiding (BasicAuth)
 
+import Servant.Auth
 import Servant.Auth.Server.Internal.Types
 import Servant.Auth.Server.Internal.ConfigTypes
 import Servant.Auth.Server.Internal.BasicAuth
 import Servant.Auth.Server.Internal.Cookie
 import Servant.Auth.Server.Internal.JWT
 
+-- | The result of running the auths 'as'
+data AuthResult auths v
+  = Authenticated v
+  | AuthFailed (ErrorList (AuthsToErrors auths))
+  deriving (Generic, Functor, Traversable, Foldable)
+
+deriving instance (Eq v, Eq (ErrorList (AuthsToErrors auths)))
+  => Eq (AuthResult auths v)
+deriving instance (Show v, Show (ErrorList (AuthsToErrors auths)))
+  => Show (AuthResult auths v)
+
+
+-- | Open type family
+type family AuthsToErrors (auths :: [*]) = (errs :: [*]) | errs -> auths where
+  AuthsToErrors '[] = '[]
+  AuthsToErrors (a ': auths) = AuthError a ': AuthsToErrors auths
+
+-- | HList as a GADT. Named differently to prevent clash with servant's HList
+data ErrorList (as :: [*]) where
+  ENil :: ErrorList '[]
+  ECons :: a -> ErrorList as -> ErrorList (a ': as)
+
+instance Show (ErrorList '[]) where
+  show ENil = "ENil"
+instance (Show a, Show (ErrorList as)) => Show (ErrorList (a ': as)) where
+  show (ECons a as) = show a ++ " ': " ++ show as
+
+instance Eq (ErrorList '[]) where
+  ENil == ENil = True
+instance (Eq a, Eq (ErrorList as)) => Eq (ErrorList (a ': as)) where
+  ECons a as == ECons a' as' = a == a' && as == as'
+
 -- | @IsAuth a ctx v@ indicates that @a@ is an auth type that expects all
 -- elements of @ctx@ to be the in the Context and whose authentication check
 -- returns an @AuthCheck v@.
-class IsAuth a v  where
+class IsAuth a v where
+  type family AuthError a = (err :: *) | err -> a
```

Comment:
Is an associated data family better here? It also has had GHC support for longer.

Reply:
I suspect that a data family is the best route, however I can't decide how to define the instances. The reason I kept the injective type family is so the error type and corresponding verification functions can be kept in their modules (e.g. `JWT`). Moving the instances to the modules causes them to be orphans (since the auth types are defined in `servant-auth`), but moving the `AuthError` types to the `Class` module mandates moving the functions as well, which I don't think is worthwhile. Any ideas here?